server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterBalanceStats.java

```diff
@@ -50,28 +53,55 @@ public static ClusterBalanceStats createFrom(
             if (dataRoles.isEmpty()) {
                 continue;
             }
-            var nodeStats = NodeBalanceStats.createFrom(routingNode, clusterState.metadata(), clusterInfo, writeLoadForecaster);
+            var nodeStats = NodeBalanceStats.createFrom(
+                routingNode,
+                clusterState.metadata(),
+                desiredBalance,
+                clusterInfo,
+                writeLoadForecaster
+            );
             nodes.put(routingNode.node().getName(), nodeStats);
             for (DiscoveryNodeRole role : dataRoles) {
                 tierToNodeStats.computeIfAbsent(role.roleName(), ignored -> new ArrayList<>()).add(nodeStats);
             }
         }
-        return new ClusterBalanceStats(Maps.transformValues(tierToNodeStats, TierBalanceStats::createFrom), nodes);
+        return new ClusterBalanceStats(
+            nodes.values().stream().mapToInt(NodeBalanceStats::shards).sum(),
+            nodes.values().stream().mapToInt(NodeBalanceStats::undesiredShardAllocations).sum(),
+            Maps.transformValues(tierToNodeStats, TierBalanceStats::createFrom),
+            nodes
+        );
     }
 
     public static ClusterBalanceStats readFrom(StreamInput in) throws IOException {
-        return new ClusterBalanceStats(in.readImmutableMap(TierBalanceStats::readFrom), in.readImmutableMap(NodeBalanceStats::readFrom));
+        return new ClusterBalanceStats(
+            in.getTransportVersion().onOrAfter(TransportVersions.UNDESIRED_SHARD_ALLOCATIONS_COUNT_ADDED) ? in.readVInt() : -1,
```

Comment:
I guess it's not too important to avoid reporting `-1` for the new fields during mix cluster scenarios? I wonder if we should use an `OptionalInt` instead?

Reply:
Oh, I see. This one is skipped with `// TEST[skip:Can't reliably test desired balance]`

Reply:
Not sure, I do not have strong preference around it